File: Asymmetric Cryptography/ECC/ECC_C/ECC.cpp
#include <iostream>
#include <stdexcept>
#include <tuple>
#include <random>
#include <cassert>

class ECC {
public:
    using Point = std::tuple<unsigned long long, unsigned long long>;

    ECC(unsigned long long p, unsigned long long a, unsigned long long b, Point G, unsigned long long n)
        : p(p), a(a), b(b), G(G), n(n) {
        // Validate curve equation
        if ((4 * a * a * a + 27 * b * b) % p == 0) {
            throw std::invalid_argument("Invalid curve: 4a^3 + 27b^2 must not be congruent to 0 mod p.");
        }

        // Validate generator point
        if (!isOnCurve(G)) {
            throw std::invalid_argument("The generator point G is not on the curve.");
        }
    }

    bool isOnCurve(Point P) const {
        auto [x, y] = P;
        return (y * y - (x * x * x + a * x + b)) % p == 0;
    }

    Point pointAddition(Point P, Point Q) const {
        if (!isOnCurve(P) || !isOnCurve(Q)) {
            throw std::invalid_argument("Both points must lie on the curve.");
        }

        auto [x1, y1] = P;
        auto [x2, y2] = Q;

        if (P == Q) {
            return pointDoubling(P);
        }

        if (x1 == x2 && y1 != y2) {
            return Point{0, 0};  // Point at infinity
        }

        if (x1 == x2) {
            return pointDoubling(P);
        }

        // Calculate the slope (lambda)
        unsigned long long slope = ((y2 - y1) * modInverse(x2 - x1, p)) % p;

        // Calculate x3 and y3
        unsigned long long x3 = (slope * slope - x1 - x2) % p;
        unsigned long long y3 = (slope * (x1 - x3) - y1) % p;

        return Point{x3, y3};
    }

    Point pointDoubling(Point P) const {
        if (!isOnCurve(P)) {
            throw std::invalid_argument("The point must lie on the curve.");
        }

        auto [x, y] = P;

        if (y == 0) {
            return Point{0, 0};  // Point at infinity
        }

        // Calculate the slope (lambda)
        unsigned long long slope = ((3 * x * x + a) * modInverse(2 * y, p)) % p;

        // Calculate x3 and y3
        unsigned long long x3 = (slope * slope - 2 * x) % p;
        unsigned long long y3 = (slope * (x - x3) - y) % p;

        return Point{x3, y3};
    }

    Point scalarMultiplication(unsigned long long k, Point P) const {
        if (k % n == 0 || !isOnCurve(P)) {
            return Point{0, 0};  // Point at infinity
        }

        Point result{0, 0};  // Point at infinity
        Point addend = P;

        while (k) {
            if (k & 1) {
                result = pointAddition(result, addend);
            }
            addend = pointDoubling(addend);
            k >>= 1;
        }

        return result;
    }

    unsigned long long generatePrivateKey() const {
        std::random_device rd;
        std::mt19937_64 eng(rd());
        std::uniform_int_distribution<unsigned long long> distr(1, n - 1);

        return distr(eng);
    }

    Point generatePublicKey(unsigned long long privateKey) const {
        if (privateKey <= 0 || privateKey >= n) {
            throw std::invalid_argument("Private key must be in the range [1, n-1].");
        }

        return scalarMultiplication(privateKey, G);
    }

private:
    unsigned long long p, a, b, n;
    Point G;

    unsigned long long modInverse(unsigned long long a, unsigned long long p) const {
        // Using Extended Euclidean Algorithm to find modular inverse
        unsigned long long m0 = p, t, q;
        unsigned long long x0 = 0, x1 = 1;

        if (p == 1)
            return 0;

        while (a > 1) {
            q = a / p;
            t = p;
            p = a % p;
            a = t;
            t = x0;
            x0 = x1 - q * x0;
            x1 = t;
        }

        if (x1 < 0)
            x1 += m0;

        return x1;
    }
};

int main() {
    // Example parameters for secp256k1 (used in Bitcoin)
    unsigned long long p = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F;
    unsigned long long a = 0;
    unsigned long long b = 7;
    ECC::Point G = {0x79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798, 
                    0x483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8};
    unsigned long long n = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141;

    // Instantiate the ECC system
    ECC ecc(p, a, b, G, n);

    // Generate keys
    unsigned long long privateKey = ecc.generatePrivateKey();
    ECC::Point publicKey = ecc.generatePublicKey(privateKey);

    std::cout << "Private Key: " << privateKey << std::endl;
    std::cout << "Public Key: (" << std::get<0>(publicKey) << ", " << std::get<1>(publicKey) << ")" << std::endl;

    return 0;
}

File: Asymmetric Cryptography/ECC/ECC_python/ECC.py
import secrets

class ECC:
    def __init__(self, p, a, b, G, n):
        """
        Initialize the ECC system.
        
        :param p: Prime number (finite field)
        :param a: Coefficient a of the elliptic curve
        :param b: Coefficient b of the elliptic curve
        :param G: Generator point (Gx, Gy)
        :param n: Order of the base point G
        """
        # Validate input types
        if not all(isinstance(i, int) for i in [p, a, b, n]):
            raise ValueError("p, a, b, and n must be integers.")
        if not isinstance(G, tuple) or len(G) != 2 or not all(isinstance(i, int) for i in G):
            raise ValueError("G must be a tuple of two integers.")
        
        self.p = p
        self.a = a
        self.b = b
        self.G = G
        self.n = n
        
        # Validate curve equation
        if (4 * a**3 + 27 * b**2) % p == 0:
            raise ValueError("Invalid curve: 4a^3 + 27b^2 must not be congruent to 0 mod p.")
        
        # Validate generator point
        if not self.is_on_curve(G):
            raise ValueError("The generator point G is not on the curve.")
    
    def is_on_curve(self, P):
        """
        Check if the point P lies on the elliptic curve.
        
        :param P: A point (x, y)
        :return: True if P lies on the curve, False otherwise
        """
        x, y = P
        return (y**2 - x**3 - self.a * x - self.b) % self.p == 0
    
    def point_addition(self, P, Q):
        """
        Perform point addition on the elliptic curve.
        
        :param P: Point P (x1, y1)
        :param Q: Point Q (x2, y2)
        :return: Resulting point R (x3, y3) from P + Q
        """
        if not self.is_on_curve(P) or not self.is_on_curve(Q):
            raise ValueError("Both points must lie on the curve.")
        
        if P == Q:
            return self.point_doubling(P)
        
        if P == (None, None):
            return Q
        if Q == (None, None):
            return P
        
        x1, y1 = P
        x2, y2 = Q
        
        if x1 == x2 and y1 != y2:
            return (None, None)  # Point at infinity
        
        if x1 == x2:
            return self.point_doubling(P)
        
        # Calculate the slope (lambda)
        slope = ((y2 - y1) * pow(x2 - x1, -1, self.p)) % self.p
        
        # Calculate x3 and y3
        x3 = (slope**2 - x1 - x2) % self.p
        y3 = (slope * (x1 - x3) - y1) % self.p
        
        return (x3, y3)
    
    def point_doubling(self, P):
        """
        Perform point doubling on the elliptic curve.
        
        :param P: Point P (x, y)
        :return: Resulting point R (x', y') from 2P
        """
        if not self.is_on_curve(P):
            raise ValueError("The point must lie on the curve.")
        
        x, y = P
        
        if y == 0:
            return (None, None)  # Point at infinity
        
        # Calculate the slope (lambda)
        slope = ((3 * x**2 + self.a) * pow(2 * y, -1, self.p)) % self.p
        
        # Calculate x' and y'
        x3 = (slope**2 - 2 * x) % self.p
        y3 = (slope * (x - x3) - y) % self.p
        
        return (x3, y3)
    
    def scalar_multiplication(self, k, P):
        """
        Perform scalar multiplication k * P on the elliptic curve.
        
        :param k: Scalar multiplier
        :param P: Point P (x, y)
        :return: Resulting point Q (x', y') from k * P
        """
        if not isinstance(k, int):
            raise ValueError("The scalar multiplier k must be an integer.")
        
        if k % self.n == 0 or not self.is_on_curve(P):
            return (None, None)  # Point at infinity
        
        result = (None, None)  # Point at infinity
        addend = P
        
        while k:
            if k & 1:
                result = self.point_addition(result, addend)
            addend = self.point_doubling(addend)
            k >>= 1
        
        return result
    
    def generate_private_key(self):
        """
        Generate a private key (a random integer).
        
        :return: Private key d
        """
        return secrets.randbelow(self.n)
    
    def generate_public_key(self, private_key):
        """
        Generate a public key corresponding to a given private key.
        
        :param private_key: The private key d
        :return: Public key Q (x, y)
        """
        if not isinstance(private_key, int):
            raise ValueError("Private key must be an integer.")
        
        if private_key <= 0 or private_key >= self.n:
            raise ValueError("Private key must be in the range [1, n-1].")
        
        return self.scalar_multiplication(private_key, self.G)

# Example parameters for secp256k1 (used in Bitcoin)
p = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F
a = 0
b = 7
G = (0x79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798, 
     0x483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8)
n = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141

# Instantiate the ECC system
ecc = ECC(p, a, b, G, n)

# Generate keys
private_key = ecc.generate_private_key()
public_key = ecc.generate_public_key(private_key)

print("Private Key:", private_key)
print("Public Key:", public_key)